### ISSUE TITLE
`#undef macro` in preamble is incorrectly overridden by `-Dmacro` in compile commands

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#1396]
### Discussed in https://github.com/clangd/clangd/discussions/1395

<div type='discussions-op-text'>

<sup>Originally posted by **han-shao-jun** November 28, 2022</sup>
Hi！I met with difficulties in my project. I would be grateful if someone could answer my hunger question.
## Problem description
My project is edited with vscode and managed with cmake. The following is my CMakeLists.txt file, which adds a predefined macro called debug. 
```cmake
cmake_minimum_required(VERSION 3.22)
project(test C)

set(CMAKE_C_STANDARD 99)
add_compile_definitions(debug)

add_executable(test main.c)

```
This is my compile_ commands. Json file. 
```json
[
{
  "directory": "D:/code/c/test/build",
  "command": "D:\\Programs\\msys64\\clang64\\bin\\clang.exe -Ddebug -g -std=gnu99 -o CMakeFiles\\test.dir\\main.c.obj -c D:\\code\\c\\test\\main.c",
  "file": "D:\\code\\c\\test\\main.c"
}
]
```
Using the following code, clangd will report incorrect syntax, but in fact, there is no syntax error, because I can compile and run.
```c
#include <stdio.h>

#undef debug
int a = 0;

#ifndef debug
int i = 0;
#endif

int main()
{
    printf("i = %d", i);
    return 0;
}


```

## System information
Output of clangd --version
```shell
$ clangd --version
clangd version 15.0.3
Features: windows
Platform: x86_64-w64-windows-gnu
```
OS: windows11
</div>

**Comment by HighCommander4** [comment]
As mentioned in the discussion, this seems to be an incorrect diagnostic arising from clangd's preamble optimization. Since the `#undef` is part of the file's preamble, it seems not to take effect in the main-file part of the file.

**Comment by HighCommander4** [comment]
Somewhat similar to #1167 but seems more impactful as it can cause false positive error diagnostics, not just warnings.

**Comment by kadircet** [comment]
i don't think this is a clangd bug, but rather a clang bug (might also be a feature). as clangd doesn't do anything specific for re-ordering PP-directives from preamble vs from the command line. i.e. if you re-write your example as:

`a.h`:
```
#pragma once
#undef debug
static int a;
```

`a.cc`:
```
#include "a.h"
#ifndef debug
int i;
#endif
i = a;
```

afterwards run:
```
clang -xc++ -Xclang -emit-pch a.h -c -o a.h.pch
clang -xc++ -Ddebug -include-pch a.h.pch -fsyntax-only a.cc # says undeclared identifier
clang -xc++ -Ddebug -fsyntax-only a.cc # works just fine
```

nice thing about this example is it works in clangd :P (just because we put both the undef and following ifdef into the same consecutive block).
but it looks like clang first loads the preamble and then applies the builtin buffer. it's not easy to reason about from my side though. because one can actually have implicit includes/pchs and the ordering of those with respect to command line flags aren't well defined i suppose.

clangd is a special case here to some extend (as we disable implicit PCHs coming from command line), so we always want PP state from preamble to take precedence over the state coming from command line.

so I'd rather see if the behaviour in clang was deliberate and we can't change it to be "always put builtin header first, and then process preamble" semantics. if that's the case, we might want to introduce a flag to control the behaviour here.

**Comment by HighCommander4** [comment]
See also https://github.com/llvm/llvm-project/issues/97759 for a reverse example, with a `-U` flag in the compile commands and a `#define` in the preamble.
